Experiment: Add full page client-side navigation experiment setting

SantosGuillamot opened the pull request:
- Part of #60219.

## What?

This pull request adds an **experiment** in Gutenberg -> Experiments to enable full page client-side navigation in your site using the Interactivity API.

**It enables the user experience [showcased in the State Of The Word](https://youtu.be/c7M4mBVgP3Y?t=1659) and in reflected in [the Movies demo](https://wpmovies.dev/)**.

[![State Of The Word 2023](https://img.youtube.com/vi/c7M4mBVgP3Y/0.jpg)](https://www.youtube.com/watch?v=c7M4mBVgP3Y?t=1659s)

Client-side navigation refers to the process of navigating between different pages without the need for a full page refresh. The Interactivity API takes care of fetching the new content from the server and replacing the relevant parts in the DOM, while maintaining the UI state. This approach provides a SPA experience, which is smoother and opens up new possibilities.

One of the limitations of this approach is that all the scripts must be compatible with the Interactivity API, and that's why it can only be enabled in controlled environments until the API is a bit more mature. Additionally, we need to explore deeper how to handle new scripts and styles.

We can differentiate between two types of client-side navigation:

### Region based client-side navigation (existing router implementation)

This type of navigation was included as a first step and it just replaces specific regions of the DOM. This is safer because just the selected regions must be compatible with the Interactivity API. This is used, for example, to [enhance the pagination of the Query Loop block](https://github.com/WordPress/gutenberg/pull/53812).

### Full page client-side navigation (enabled by this experiment)

This type of navigation does replace the whole DOM and not just specific regions. As mentioned above, there might be incompatibilities with other scripts, so it should only be used in controlled environments. This experiment should help to clarify what is needed and how it could be implemented.

## Why?
The idea is to include it as experimental to start testing it and explore what needs to be done to offer it as a site option without the experimental flag.

## How?
There are a few changes in this pull request:
* It adds a new setting in the experiments panel: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-e6bd94e1847aea4b70a6cc1ac9b2471dd3d882f6cbf1ab51438e23b02ca5194fR130-R141). When that option is enabled, it loads conditionally the relevant PHP code.
* If the experiment is enabled, it adds `data-wp-interactive` to all the root blocks: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-4c31288ce2b84520351c4319e32cbb71215b562d80132a3f7b5441a89bd8795fR31-R72).
* If the experiment is enabled, it registers a new interactivity config setting that can be read in PHP or in JS: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-4c31288ce2b84520351c4319e32cbb71215b562d80132a3f7b5441a89bd8795fR6-R7).
* It creates a new `interactivity-router` module named `full-csn` to handle the whole site and not only regions: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-c6f39cecb38b1631b9cfa08afbfefd80805da6e69fc2b8eea4868c911dc601e1R1-R205). It is based on the one there was in the initial experiments: [link](https://github.com/WordPress/block-interactivity-experiments/blob/main-wp-directives-plugin/src/runtime/router.js).
* If the experiment is enabled, it registers the new router and, if not, it loads the old one like it is doing in trunk: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-2061d83c07f2215ab0526ac98b601eb910aabdc7a20ac8487be836f54287b31bR24-R39).
* If the experiment is enabled, it adds some directives to all links to prefetch the new page and navigate on click using the router: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-4c31288ce2b84520351c4319e32cbb71215b562d80132a3f7b5441a89bd8795fR9-R27). This should probably not be done by Gutenberg but each site owner should decide where to use it. It is useful now for testing.
* It adapts the query loop block to the new router and loads the script always when the experiment is enabled.

## Testing Instructions
1. Enable the experiment in Gutenberg -> Experiments.
2. Click on links in your site and check that the page is not refreshed.

SantosGuillamot committed:
Add Gutenberg experiment option

SantosGuillamot committed:
Add config option and directives in PHP

SantosGuillamot committed:
Load full CSN logic conditionally

SantosGuillamot committed:
Add `data-wp-interactive` root

SantosGuillamot committed:
Change variables names

SantosGuillamot committed:
Register different scripts if the experiment is enabled

SantosGuillamot committed:
Require experimental code once interactivity is loaded

SantosGuillamot committed:
Change experiment namespace

SantosGuillamot committed:
Move full-csn logic to interactivity-router

SantosGuillamot committed:
Add proper support for prefetch

SantosGuillamot committed:
Adapt query loop

SantosGuillamot committed:
Fix modules error after csn

SantosGuillamot committed:
Add initial page to cache

SantosGuillamot committed:
WIP: Fix scripts loading after csn

SantosGuillamot committed:
Simplify code

SantosGuillamot committed:
Adapt query loop block

SantosGuillamot committed:
Fix full CSN when queryID is not defined

SantosGuillamot committed:
Remove preload logic

SantosGuillamot committed:
Change full csn conditional in query

SantosGuillamot committed:
Use only one app in the body

SantosGuillamot committed:
Use getRegionRootFragment and initialVdom

SantosGuillamot committed:
Adapt all query loop blocks

SantosGuillamot committed:
Add key to query loop block

SantosGuillamot committed:
Add `yield` to query block actions

c4rl0sbr4v0 commented:
To check:
 - If I click on the Site title after a reload in a different page. The URL is not being updated.

cbravobernal commented on lib/experimental/full-client-side-navigation.php:
```diff
@@ -0,0 +1,31 @@
+	}
+	// Hack to add the necessary directives to the body tag.
+	// TODO: Find a proper way to add directives to the body tag.
+	return (string) $p . '<body data-wp-interactive="core/experimental" data-wp-context="{}">';
```
Is this adding a body on each block?

SantosGuillamot replied:
Technically, yes, although it translates into just having the unique body with the directive.

That's why I added a note that it is a hack, and we need to find a proper way to add the directive there. I couldn't find one proper way yet.

cbravobernal commented on lib/interactivity-api.php:
```diff
@@ -21,12 +21,22 @@ function gutenberg_reregister_interactivity_script_modules() {
+		wp_enqueue_script_module( '@wordpress/interactivity-router' );
+	} else {
+		wp_register_script_module(
+			'@wordpress/interactivity-router',
```
Calling a different script with the same name can cause misunderstandings if people use or extend the interactivity-router module. 
As is an experiment, we can keep going, but we will need to integrate csn into the library one received some feedback and testing.

SantosGuillamot replied:
That makes sense 🙂 We should probably need to look for a better way once it is not an experiment.

gziolo replied:
That's confusing to have two different files to play the same purpose. Would it be possible to have the functionality related to full client-side navigation in the existing router file but guarded with `if ( process.env.IS_GUTENBERG_PLUGIN ) {` check like in all other packages? That would simplify everything and better align with the rest of packages.

SantosGuillamot replied:
> Would it be possible to have the functionality related to full client-side navigation in the existing router file but guarded with if ( process.env.IS_GUTENBERG_PLUGIN ) { check like in all other packages?

As this is an experiment, we wanted to avoid sending extra JS that is not needed in the current implementation.

Anyway, I'll think how this can be improved.

gziolo replied:
Webpack, when it sees `process.env.IS_GUTENBERG_PLUGIN` set to `false` removes all the code behind this check. In effect, it won't impact the bundle size at all in WordPress core. In shouldn't matter that much in the Gutenberg plugin while running the experiment.

SantosGuillamot replied:
Oh, I didn't know that 🙁 I can give that a try then. Although I am not sure yet how it should work because the conditional should not be when is Gutenberg but when the experiment is enabled.

gziolo replied:
What is the difference in the size of the bundle? The question is whether it’s worth complicating the entire build system for a single experiment.

SantosGuillamot replied:
From a quick test, it seems the router would increase from 1,7 kb to 3.2 kb, more or less.

When you say "complicating the entire build system" do you mean the PHP conditional used here or creating a new entry in the interactivity webpack: [link](https://github.com/WordPress/gutenberg/pull/59707/files#diff-3d1911952fc568c4108afcd3bee2a8a5c5aea693d240df8a1a2812337256374dR22).

I believe we can look for alternatives to avoid both.

gziolo replied:
It doesn't seem like a huge problem then given that the router is loaded asynchronously only when it's used. In effect, a user needs to hover the mouse over the link or click on it:

<img width="803" alt="Screenshot 2024-03-20 at 14 32 06" src="https://github.com/WordPress/gutenberg/assets/699132/399afd44-308b-4499-bfbd-e62605c4769d">


<img width="985" alt="Screenshot 2024-03-20 at 14 32 51" src="https://github.com/WordPress/gutenberg/assets/699132/44ffed1a-48e6-4f7c-a43b-995eb074b190">

SantosGuillamot replied:
I've pushed another option I was considering:
* Keep the current scripts as they are: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/652a2a2de432dd132c364296565d99b5ca2df4a6).
* Register and load a new script for full client-side navigation when the experiment is enabled: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/442248c809b4eebffae92fbc1120a4a164f069ce).
* Load the original `@wordpress/interactivity-router` in the query loop only when the experiment is not enabled: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/0f3bc44fbe754f5cadd0957c93505078bfe7f22a).

@DAreRodz It'd be great to know your opinion about the best approach here.

DAreRodz replied:
Huh, this one is tricky. Here's my two cents.

A pattern we thought about is that modules can define a light interface that is implemented by other async modules.

```jsx
import { store } from '@wordpress/interactivity';

store( 'somePlugin', {
	actions: {
		*doSomething() {
			// Let's assume `some-plugin/implementation` is a heavy module.
			yield import( 'some-plugin/implementation' );
			yield store( 'somePlugin/implementation' ).actions.doSomething();
		}
	}
} );
```

That approach doesn’t strictly apply here (we are already loading the `@wordpress/interactivity-router` module asynchronously), but my initial idea was to use that pattern to conditionally load the *region-based* or the *full-page* implementation of the router from `@wordpress/interactivity-router`, based on the `fullClientSideNavigation` config prop.

The immediate benefit is that blocks calling `store( 'core/router' ).actions.navigate` don’t have to be aware of which implementation should be used, or which module they have to import (see https://github.com/WordPress/gutenberg/pull/59707/commits/0f3bc44fbe754f5cadd0957c93505078bfe7f22a).

One side effect of this approach is that it would require defining two extra modules (two more entry points in Webpack and two more modules to register/enqueue in WP). I guess that’s *acceptable*?

When I discussed this with @SantosGuillamot a while ago, we saw that it was easier to just replace the module we send to the client when the `@wordpress/interactivity-router` is requested. That’s why we did so, although that’s probably not a good idea, as @gziolo mentioned. 😅

---

Anyways, as this is an experiment, I’m fine with any approach *for now*. We can iterate later on to find the best patterns for this use case.

SantosGuillamot replied:
I don't mind implementing any approach, but it'd be great to make a decision so we can unblock this PR. We have to keep in mind that this is an experiment, so how the different modules will be handled can be explored deeper. The idea of having an experiment was precisely having it available without worrying too much about those aspects.

So it seems there are mainly three options at this point:

1. Include the conditional logic in the existing router

This increases the bundle size of the router even if the experiment is not enabled. The router is only loaded right now when the user hovers over a link in the query loop block.

On the other hand, it doesn't require changes in the webpack configuration.

2. Existing solution in this PR: Create two different routers, the existing one and a new one for full page client-side navigation

This requires creating a new module in WebPack, which is loaded when the experiment is enabled. When the experiment is enabled, it doesn't load the existing router for the query loop.

3. Explore the approach suggested by David [here](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1535479413)

I am not 100% sure how it differs from the solution implemented in this PR, but I could explore it more deeply.

---

@DAreRodz @gziolo Any strong opinions about the correct approach to unblock this experiment?

gziolo replied:
The most important aspect, from my perspective, would be ensuring that the code doesn't diverge between two files. In particular, when a bug fix gets applied for the existing router, it's going to be very hard to ensure it gets mirrored in the other file. In effect, the simplest approach to unblock it here would be to move common logic to another file and import it from both entry points:

```
|-- interactivity-router
    |-- src
        |-- index.js // old entry point
        |-- experimental.js // experimental entry point
        |-- shared.js // shared logic, but can be also several files
```

SantosGuillamot replied:
After taking a deeper look at it, I agree that most of the code of the router should be reused and having just one router checking the `navigationMode` seems the right approach to me. I've changed the PR to reflect that. Let me know what you think.

I've checked the bundle size and it increases from 1.7kB to 3.1kB. However, the code that handles the head tags is new and it will be needed for both modes in the future. Right now it is only used in this experiment but when it is tested and more mature, it should be used for region based client-side navigation as well.

EDIT: I have to review the implementation though. I believe something is not working as expected. But you can get an idea of how it could look like.

cbravobernal commented on packages/interactivity-router/src/full-csn.js:
```diff
@@ -0,0 +1,200 @@
+	! event.ctrlKey && // Open in new tab (Windows).
+	! event.altKey && // Download.
+	! event.shiftKey &&
+	! event.defaultPrevented;
```
Is this variable covering a user can open a link in a new tab by pressing the mouse wheel?

SantosGuillamot replied:
I've tested it, and it seems to be working. From my tests, it seems that when you press the mouse wheel, it doesn't trigger an onClick event, so `actions.navigate` is not called, and it just works as expected.

cbravobernal reviewed (approved):
I couldn't find anything that could block this PR to be merged as an experiment.
All the comments can be handled in future PRs.

gziolo reviewed (changes requested):
Can you provide the full context for this experiment so people can learn what they will be opting into when these changes land. Ideally, it should have a related issue that gives more in-depth explanation about the feature proposed.

gziolo commented on tools/webpack/interactivity.js:
```diff
@@ -19,6 +19,7 @@ module.exports = {
 	entry: {
 		index: './packages/interactivity',
 		router: './packages/interactivity-router',
+		'full-csn': './packages/interactivity-router/src/full-csn.js',
```
Let's avoid using acronyms like `csn` in the code that aren't easy to unpack when you don't have the full context. A quick Google search doesn't show anything related for [CSN](https://acronyms.thefreedictionary.com/CSN).

SantosGuillamot replied:
In [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/cd7a112cf072e2ce3d49c2ab912cbdc19eac77eb), I renamed everything to use "Full page client-side navigation" and "full page router". I hope that with that, summed to the new explanation in the opening comment, it's enough for the experiment. Let me know if you feel it isn't clear yet.

gziolo commented on packages/block-library/src/post-template/index.php:
```diff
@@ -45,9 +45,13 @@ function block_core_post_template_uses_featured_image( $inner_blocks ) {
+	$page_key                        = isset( $block->context['queryId'] ) ? 'query-' . $block->context['queryId'] . '-page' : 'query-page';
+	$enhanced_pagination             = isset( $block->context['enhancedPagination'] ) && $block->context['enhancedPagination'];
+	$interactivity_api_router_config = wp_interactivity_config( 'core/router' );
+	if ( ! empty( $interactivity_api_router_config ) && ! empty( $interactivity_api_router_config['fullClientSideNavigation'] ) ) {
```
`empty` is a language construct and it performs internally all necessary checks:
```suggestion
	if ( ! empty( $interactivity_api_router_config['fullClientSideNavigation'] ) ) {
```

The same applies to all other occurrences.

SantosGuillamot replied:
Changed in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/a26bfa70e582d5b06d186394301a9748db8a90eb).

gziolo commented on packages/block-library/src/query-pagination-next/index.php:
```diff
@@ -15,10 +15,14 @@
+	$page_key                        = isset( $block->context['queryId'] ) ? 'query-' . $block->context['queryId'] . '-page' : 'query-page';
+	$enhanced_pagination             = isset( $block->context['enhancedPagination'] ) && $block->context['enhancedPagination'];
+	$interactivity_api_router_config = wp_interactivity_config( 'core/router' );
+	if ( ! empty( $interactivity_api_router_config ) && ! empty( $interactivity_api_router_config['fullClientSideNavigation'] ) ) {
```
If the setting enforces enhanced pagination, should it also impact the UI in the block editor? What is the plan regarding the Query Loop setting - force page reload.

Edit: it looks like the setting gets disabled

gziolo commented on packages/block-library/src/query/index.php:
```diff
@@ -51,8 +55,8 @@ function render_block_core_query( $attributes, $content, $block ) {
 			$p->set_attribute( 'data-wp-router-region', 'query-' . $attributes['queryId'] );
-			$p->set_attribute( 'data-wp-init', 'callbacks.setQueryRef' );
 			$p->set_attribute( 'data-wp-context', '{}' );
+			$p->set_attribute( 'data-wp-key', $attributes['queryId'] );
```
I'm also wondering about this change that gets applied when the experiment is both enabled and disabled. Was `setQueryRef` in use before? I can't find it in the source code.

SantosGuillamot replied:
From what I understood from @DAreRodz , it was part of an old implementation, and it seems it wasn't removed. I couldn't find it either.

It seems the `setQueryRef` callback:

* Was added [here](https://github.com/WordPress/gutenberg/commit/19b2bf9cc51074e3bcc9448689618912c391ab5d).
* The JS logic was removed [here](https://github.com/WordPress/gutenberg/pull/58496/files), but we didn't remove the directive.

cbravobernal replied:
There is an issue planning to unset the queryId if the query inherits its options.
https://github.com/WordPress/gutenberg/issues/60374

gziolo commented on packages/block-library/src/query/view.js:
```diff
@@ -43,25 +43,27 @@ store(
 				if ( isValidLink( ref ) ) {
-					const { actions } = yield import(
-						'@wordpress/interactivity-router'
+					yield import( '@wordpress/interactivity-router' );
```
Another place where the current functionality gets updated is when the experiment is enabled and disabled. I assume it's correct, but someone could confirm.

SantosGuillamot replied:
I believe I tested this, but I will double-check.

SantosGuillamot replied:
Okay, if I understood you correctly, I believe this work as expected.
* I start without the experiment: The query loop navigation works as in `trunk`.
* I enable the experiment: The query loop works with the logic included in this PR.
* I disable the experiment: The query loop works as in `trunk` again.

Is that what you were referring to?

gziolo replied:
Yes, with and without experiment it should work correctly as previously.

gziolo commented on lib/experimental/full-client-side-navigation.php:
```diff
@@ -0,0 +1,31 @@
+		return $content;
+	}
+	$p = new WP_HTML_Tag_Processor( $content );
+	while ( $p->next_tag( array( 'tag_name' => 'a' ) ) ) {
```
What is the expectation for the external URLs? What about links pointing to an anchor on the same page?

SantosGuillamot replied:
Those cases are currently handled by the navigate function:
* It checks if the URL is external [here](https://github.com/WordPress/gutenberg/pull/59707/files#diff-c6f39cecb38b1631b9cfa08afbfefd80805da6e69fc2b8eea4868c911dc601e1R121-R126).
* I just added some logic to scroll to the anchor if exists [here](https://github.com/WordPress/gutenberg/pull/59707/files#diff-c6f39cecb38b1631b9cfa08afbfefd80805da6e69fc2b8eea4868c911dc601e1R158-R163).

gziolo replied:
Nice, thank yiy for confirming 👍🏻 

One additional consideration, should that behavior get only applied to blocks marked as interactive through `supports.interactivity`?

SantosGuillamot replied:
Mmm not sure if for the experiment we should apply it only to interactive blocks. For example, we might want to add full page client-side navigation to images or links included in the footer, which are not interactive.

SantosGuillamot committed:
Revert conditional scripts depending on the experiment

SantosGuillamot committed:
Register `interactivity-router-full-client-side-navigation` in the experiment

SantosGuillamot committed:
Load router conditionally in query loop

SantosGuillamot committed:
Scroll to anchor

SantosGuillamot committed:
Remove unnecessary empty conditional

SantosGuillamot committed:
Fix back and forward buttons

SantosGuillamot committed:
Fix query loop

SantosGuillamot committed:
Remove unnecessary conditional

gziolo commented on packages/interactivity-router/src/full-csn.js:
```diff
@@ -0,0 +1,205 @@
+				}
+
+				// Scroll to the anchor if exits in the link.
+				if ( !! event.target?.hash ) {
```
That's a nice improvement 💯

SantosGuillamot commented:
> Can you provide the full context for this experiment so people can learn what they will be opting into when these changes land. Ideally, it should have a related issue that gives more in-depth explanation about the feature proposed.

I have updated the opening comment, although I agree client-side navigation should have a deeper explanation to discuss all the details as it is a complex topic. Let me know if that's enough for now or I should work on something else.

SantosGuillamot committed:
Use full page client-side navigation naming

SantosGuillamot committed:
Change comments

SantosGuillamot committed:
Use render_block_data to change query attribute

SantosGuillamot committed:
Refactor JavaScript logic

SantosGuillamot committed:
Remove unused variable

SantosGuillamot committed:
Revert changes in query block view.js file

SantosGuillamot committed:
Remove unnecessary export from interactivity

SantosGuillamot committed:
Move logic to the existing router

SantosGuillamot committed:
Use vdom.get document.body

SantosGuillamot committed:
Remove nextTick function

SantosGuillamot committed:
Only call getElement when it is an event

SantosGuillamot committed:
Allow instanceof URL in navigate

gziolo commented on packages/interactivity-router/src/index.js:
```diff
@@ -128,23 +191,30 @@ export const { state, actions } = store( 'core/router', {
-		 * @param {number}  [options.timeout]                  Time until the navigation is aborted, in milliseconds. Default is 10000.
-		 * @param {boolean} [options.loadingAnimation]         Whether an animation should be shown while navigating. Default to `true`.
-		 * @param {boolean} [options.screenReaderAnnouncement] Whether a message for screen readers should be announced while navigating. Default to `true`.
+		 * @param {string|Object} eventOrUrl                         The page href or the event handler in case it is used directly in a directive.
```
There is type-checking included later which defines the exact type:
```suggestion
		 * @param {string| Event} eventOrUrl                         The page href or the event handler in case it is used directly in a directive.
```

SantosGuillamot replied:
Changed as part of [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/f7e5212961ad3909abc3b6665f4a1f06bc66a3f6).

gziolo commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,47 @@
+// Set the navigation mode to full page client-side navigation.
+wp_interactivity_config( 'core/router', array( 'navigationMode' => 'fullPage' ) );
+
+// Enqueue the interactivity router script.
```
I don't think that `init` is the correct hook because this script shouldn't get enqueued in WP Admin. 

Nitpick: for function signatures, PHPDoc format should be used:
```suggestion
/**
 * Enqueues the interactivity router script.
 */
```

gziolo replied:
[`wp_enqueue_scripts`](https://developer.wordpress.org/reference/hooks/wp_enqueue_scripts/) seems like a proper hook to use here.

SantosGuillamot replied:
Done in this commit: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/181614e418fa958f90077d7a3e9a409326126299).

gziolo commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,47 @@
+
+add_action( 'init', '_gutenberg_enqueue_interactivity_router', 20 );
+
+// Add directives to all links.
```
Same as above, PHPDoc should be used and the params and return values should get documented.

SantosGuillamot replied:
Done as part of [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/36e090173ee052f9d5bbb8a92eca3304e2df2bcb).

gziolo commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,47 @@
+	return $parsed_block;
+}
+
+add_filter( 'render_block_data', '_gutenberg_add_enhanced_pagination', 10 );
```
Nitpick: 10 is the default priority, so it can be omitted.

SantosGuillamot replied:
Done as part of [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/36e090173ee052f9d5bbb8a92eca3304e2df2bcb).

SantosGuillamot committed:
Fix full page client-side navigation

gziolo commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,47 @@
+}
+
+// TODO: Explore moving this to the server directive processing.
+add_filter( 'render_block', '_gutenberg_add_client_side_navigation_directives', 10, 2 );
```
There is only one param consumed by the callback, so last 2 params are equal to default settings.

SantosGuillamot replied:
Done as part of [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/36e090173ee052f9d5bbb8a92eca3304e2df2bcb).

gziolo commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,47 @@
+ */
+
+// Set the navigation mode to full page client-side navigation.
+wp_interactivity_config( 'core/router', array( 'navigationMode' => 'fullPage' ) );
```
Should this code be colocated with enqueuing the router? It also only applies to the front end for the time being.

SantosGuillamot replied:
Done as part of [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/181614e418fa958f90077d7a3e9a409326126299).

gziolo commented on packages/interactivity-router/src/index.js:
```diff
@@ -1,7 +1,16 @@
+/**
```
Nitpick: WordPress dependencies are always listed before internal dependencies by convention: https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#imports

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/455b8d74e4b9fd2dabe3960b6d168a0791830546).

gziolo commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,77 @@
+// Function to update only the necessary tags in the head.
```
In the follow up work - a dedicated PR - it would be great to use JSDoc comments for all functions:

https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#javascript-documentation-using-jsdoc

SantosGuillamot replied:
I added som JSDoc comments to the new functions introduced by this PR in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/a2c1a070da65542dcf10d45e728c4be40010c56a). However, there are still missing docs in other interactivity API functions.

SantosGuillamot committed:
Use `wp_enqueue_scripts` hook

SantosGuillamot committed:
Clean PHP code and docs

SantosGuillamot committed:
Move internal dependencies after WordPress ones

SantosGuillamot committed:
Add initial JSDocs to head helper functions

SantosGuillamot committed:
Allow URL instance in prefetch function

gziolo commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,92 @@
+ * @param {Document} document     The document from which to fetch head assets. It should support standard DOM querying methods.
+ * @param {Map}      headElements A map of head elements to modify tracking the URLs of already processed assets to avoid duplicates.
+ *
+ * @return {void}
```
It returns an array of HTML elements, for example [`HTMLTitleElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLTitleElement) that inherits from `HTMLElement`. That probably should be:

```suggestion
 * @return {HTMLElement[]}
```

SantosGuillamot replied:
You're totally right. Copying and pasting issues 😄 I changed it in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/743926f0160e985a4641c3b158f64700e8c656ea).

gziolo commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,92 @@
+ * @async
+ * @param {Array} newHead The head elements of the new page.
+ *
+ * @return {void}
```
If there is no return value, you can skip `@return` line.

SantosGuillamot replied:
Changed in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/743926f0160e985a4641c3b158f64700e8c656ea).

SantosGuillamot committed:
Properly support prefetch

SantosGuillamot committed:
Fix JSDoc comments

gziolo commented on packages/interactivity-router/src/index.js:
```diff
@@ -102,12 +134,41 @@ window.addEventListener( 'popstate', async () => {
-// Cache the initial page using the intially parsed vDOM.
+// Initialize the router and cache the initial page using the initial vDOM.
+// Once this code is tested and more mature, the head should be updated for region based navigation as well.
+if ( navigationMode === 'fullPage' ) {
```
I would also consider wrapping all code occurrences of the code for the full page navigation with:

```js
if ( process.env.IS_GUTENBERG_PLUGIN ) {
    if ( navigationMode === 'fullPage' ) {
        // Code here.
    }
}
```

It will ensure that these changes won't get accidentally exposed in WordPress core.

SantosGuillamot replied:
Addressed in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/46ce03183471cb5379a7907f20c04e9d9d23149b).

luisherranz commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,60 @@
+	$p = new WP_HTML_Tag_Processor( $content );
+	while ( $p->next_tag( array( 'tag_name' => 'a' ) ) ) {
+		if ( empty( $p->get_attribute( 'data-wp-on--click' ) ) ) {
+			$p->set_attribute( 'data-wp-on--click', 'core/router::actions.navigate' );
```
For what it is worth, I think we should use event delegation for this. Something similar to this, but adding our checks (if the element already has a `data-wp-on--click`...): https://gist.github.com/devongovett/919dc0f06585bd88af053562fd7c41b7

DAreRodz replied:
That's a good idea. I guess it makes sense if we address that in a follow-up PR, right?

westonruter replied:
@felixarntz related to the discussion in https://github.com/WordPress/gutenberg/discussions/60574 regarding async event handlers, if event delegation were employed like this above [Gist](https://gist.github.com/devongovett/919dc0f06585bd88af053562fd7c41b7) does, then this is another area where synchronous event handlers can negatively impact INP. It would be likely that analytics plugins would adopt similar event delegation rather than having to use the tag processor to inject their own `click` event directives on each link.

SantosGuillamot replied:
For this experiment, I added [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/d72cd023eae40645e8e787f24d91f515b12ce4ee) to use `document.addEventListener` instead of the previous directives. I also added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore this more deeply.

gziolo commented:
@DAreRodz, what are new e2e tests that we could add in a follow-up PR that would cover the changed/introduced functionality?

DAreRodz commented:
> How should we proceed here? In the meantime, I'll go with option 1 to ensure all tests pass.

@SantosGuillamot, let's choose option 2. The `navigate` action is not designed to support object instances for the `href` attribute. We need to update the tests instead. As you mentioned, it's as simple as passing `window.location.href` or `url.href`.

We could change that in the future if necessary.

---

By the way, do you know why the tests are failing now and not before? 🤔

SantosGuillamot commented:
> By the way, do you know why the tests are failing now and not before? 🤔

Tests were failing in this PR because we modified the router to add a check ensuring it was a string: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/8f09d2b56437d97838516ce60916670418550a4b#diff-6b0e711ee1a7c83c535664372c8a0db6b2025cb33e9263bab915dff5145e01e7L317-L320). We can go back to the string check if we choose option 2.

DAreRodz commented:
> what are new e2e tests that we could add in a follow-up PR that would cover the changed/introduced functionality?

@gziolo, I guess the e2e tests could cover, e.g., these cases with the experiment enabled:

- A site becomes interactive without adding interactive blocks.
- Internal links navigate without reloading the page.
- External links navigate reloading the page.
- State/context is preserved/updated during navigation
- All cases from the `router-navigate` spec file should work with the experiment enabled.

DAreRodz commented:
> Tests were failing in this PR because we modified the router to add a check ensuring it was a string: [link](https://github.com/WordPress/gutenberg/pull/59707/commits/8f09d2b56437d97838516ce60916670418550a4b#diff-6b0e711ee1a7c83c535664372c8a0db6b2025cb33e9263bab915dff5145e01e7L317-L320). We can go back to the string check if we choose option 2.

I see. Yeah, let's go back to the string check. We could change the action later to accept `URL` instances if needed. In that case, I would use the `.href` property internally to prevent any problem with URL comparison, like we're doing [here](https://github.com/WordPress/gutenberg/pull/59707/commits/8f09d2b56437d97838516ce60916670418550a4b#diff-6b0e711ee1a7c83c535664372c8a0db6b2025cb33e9263bab915dff5145e01e7R258-R261).

DAreRodz commented on packages/block-library/src/query/edit/inspector-controls/enhanced-pagination-control.js:
```diff
@@ -15,9 +15,13 @@ export default function EnhancedPaginationControl( {
 	let help = __( 'Browsing between pages requires a full page reload.' );
-	if ( enhancedPagination ) {
+	if ( fullPageClientSideNavigation ) {
+		help = __( 'Full page client-side navigation enabled.' );
```
Perhaps indicating here that "full-page, client-side navigation" is an experimental setting would be good. 🤔

SantosGuillamot replied:
Changed in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/eeb6491d31d32990401b2397d10ba2eb494afa87).

DAreRodz commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,60 @@
+	}
+	// Hack to add the necessary directives to the body tag.
+	// TODO: Find a proper way to add directives to the body tag.
+	return (string) $p . '<body data-wp-interactive="core/experimental" data-wp-context="{}">';
```
We could use a static variable here to add the `<body>` tag once.

```suggestion
	static $body_interactive_added;
	if ( ! $body_interactive_added ) {
		$body_interactive_added = true;
		return (string) $p . '<body data-wp-interactive="core/experimental" data-wp-context="{}">';
	}

	return (string) $p;
```

gziolo replied:
That'd be important to ensure it gets added only once.

westonruter replied:
Is this injecting an additional `<body>` tag after the `body` has already been opened? Like this?

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
  </head>
  <body>
    <h1>
      Hello World!
    </h1>
    <body class="hey">
  </body>
</html>
```

Surprisingly, this actually works:

![image](https://github.com/WordPress/gutenberg/assets/134745/5337e3fb-1672-4683-af7a-ba9d63e69501)

Maybe I shouldn't be surprised given HTML's loose parsing rules. But I can imagine other plugins that try to parse the HTML to, for example, apply various optimizations would get might confused when encountering multiple `<body>` tags.

Shouldn't the attributes rather get injected on the existing `body` tag instead with the HTML Tag Processor? I see this is not currently possible since the `body` tag is hard-coded in [`template-canvas.php`](https://github.com/WordPress/wordpress-develop/blob/f525e665b6ea6e88289d82cad5fc1dfac57db109/src/wp-includes/template-canvas.php#L20). This brings up the question again of output buffering for the entire template (Core-43258). It's something I've been working on in the context of the Performance team's [Optimization Detective](https://wordpress.org/plugins/optimization-detective/) plugin which output buffers the entire template and then uses HTML Tag Processor to do optimizations.

SantosGuillamot replied:
I've added the check in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/57ef83ceb0b51a09e522d5bea4218c5acb755c1d) and I added a follow-up task in [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore how to do this properly.

westonruter replied:
Opened PR to address this: https://github.com/WordPress/gutenberg/pull/61212

gziolo replied:
Nice, thank you for sharing @westonruter.

DAreRodz reviewed (approved):
Well done! 👏  There are some minor changes to make, though. 👇 

The `render_block_data` hook doesn't work as expected because it doesn't modify the block attributes, just the values inside the `WP_Block`'s `parsed_block` property.

We can add the following code for now:

```patch
diff --git a/packages/block-library/src/query/index.php b/packages/block-library/src/query/index.php
index 88b3af438db..a8fb44fc1e2 100644
--- a/packages/block-library/src/query/index.php
+++ b/packages/block-library/src/query/index.php
@@ -21,6 +21,13 @@ function render_block_core_query( $attributes, $content, $block ) {
 		&& true === $attributes['enhancedPagination']
 		&& isset( $attributes['queryId'] );
 
+	if (
+		isset( $block->parsed_block['attrs']['enhancedPagination'] )
+			&& $block->parsed_block['attrs']['enhancedPagination']
+	) {
+		$is_interactive = true;
+	}
+
 	// Enqueue the script module and add the necessary directives if the block is
 	// interactive.
 	if ( $is_interactive ) {

```

Also, we would have to take another look a the `head.js` code later on and ensure it works for most cases.

michalczaplinski commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,91 @@
+ *
+ * @return {HTMLElement[]} Returns an array of HTML elements representing the head assets.
+ */
+export const fetchHeadAssets = async ( document, headElements ) => {
```
We should probably rename the `document` parameter to something else because `document` is a global.

SantosGuillamot replied:
Changed in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/4f9d728ccb8a401b4e388df6f979ed7fa2cf1bf6).

michalczaplinski commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,91 @@
+ * @param {Document} document     The document from which to fetch head assets. It should support standard DOM querying methods.
+ * @param {Map}      headElements A map of head elements to modify tracking the URLs of already processed assets to avoid duplicates.
+ *
+ * @return {HTMLElement[]} Returns an array of HTML elements representing the head assets.
+ */
```
```suggestion
 * @return {Promise<HTMLElement[]>} Returns an array of HTML elements representing the head assets.
```

michalczaplinski commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,91 @@
+ * @param {Document} document     The document from which to fetch head assets. It should support standard DOM querying methods.
+ * @param {Map}      headElements A map of head elements to modify tracking the URLs of already processed assets to avoid duplicates.
+ *
+ * @return {HTMLElement[]} Returns an array of HTML elements representing the head assets.
```
```suggestion
 * @return {Promise<HTMLElement[]>} Returns an array of HTML elements representing the head assets.
```

michalczaplinski commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,91 @@
+			Array.from( tags ).map( async ( tag ) => {
+				const attributeValue = tag.getAttribute( attribute );
+				if ( ! headElements.has( attributeValue ) ) {
+					const response = await fetch( attributeValue );
```
Would be nice to add a try/catch here to avoid any "unhandled promise rejection" errors

SantosGuillamot replied:
Added in [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/a5b9e80f495208209ae76178a7bff2d999e949c1).

michalczaplinski commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,91 @@
+
+				const headElement = headElements.get( attributeValue );
+				const element = document.createElement( tagName );
+				element.innerText = headElement.text;
```
For styles we should use `element.textContent` because `innerText` might [cause a
reflow](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#differences_from_innertext)

SantosGuillamot replied:
I added it as a follow-up to explore after this experiment.

michalczaplinski reviewed (commented):
I'm a little concerned about fetching scripts in `fetchHeadAssets()` without taking into account the script order.

Imagine this scenario. The page we are fetching has 2 scripts:

```
<head>
  <!-- 1MB, loads slowly -->
  <script src="A.js"></script>
  
  <!--  1kb, loads fast, DEPENDS on script A !!! -->
  <script src="B.js"></script>
</head>
```

Using our current approach, **script B** will get fetched first and added to the DOM, but the **script A** will not have loaded yet. This can be reasonably common for pages that e.g. enqueue some library and subsequently enqueue scripts that use that library.

----

I wonder if instead of fetching scripts using `fetch()`, getting their contents and shoving it into a new `<script>` tag we could do something else: Instead, we could insert the script into the head and add the appropriate `src` attribute and let the browser handle the loading of the script. We could still keep track of which scripts have already been loaded using the `headElements` map - we would just offload the script loading logic to the browser.

gziolo commented:
> I wonder if instead of fetching scripts using `fetch()`, getting their contents and shoving it into a new `<script>` tag we could do something else: Instead, we could insert the script into the head and add the appropriate `src` attribute and let the browser handle the loading of the script. We could still keep track of which scripts have already been loaded using the `headElements` map - we would just offload the script loading logic to the browser.

I shared that before, but this is how a similar logic is handled for blocks installed on the fly in the block editor after inserting from the Plugin Directory:

https://github.com/WordPress/gutenberg/blob/6b78b7d0b2731a6b7b3bcccf53fd2b0f55d132f7/packages/block-directory/src/store/load-assets.js#L16-L48

It would also fit here and resolve most of the issues that @michalczaplinski raised.

michalczaplinski commented:
> I'm a little concerned about fetching scripts in fetchHeadAssets() without taking into account the script order.

To be clear, I'm happy if we only address the minor issues now and try to resolve the bigger problem of the script loading order in follow-up PRs 🙂 .

michalczaplinski commented:
It would also be worth checking if https://github.com/WordPress/block-interactivity-experiments/issues/242 is still an issue or if it had been fixed.

felixarntz commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,60 @@
+			$p->set_attribute( 'data-wp-on--click', 'core/router::actions.navigate' );
+		}
+		if ( empty( $p->get_attribute( 'data-wp-on--mouseenter' ) ) ) {
+			$p->set_attribute( 'data-wp-on--mouseenter', 'core/router::actions.prefetch' );
```
Per @gziolo's comment on https://make.wordpress.org/core/2024/04/09/speculative-loading-in-wordpress/, if we eventually move forward with adopting the Speculation Rules API in WordPress core, we would need to consider whether or how this conflicts with it.

FWIW the Speculation Rules API is a web technology, and if it allows us to replace this manual implementation, that would probably be a good thing. That said, I'm not sure how it would behave here given that we don't do a "real" navigation when those links are clicked but client-side navigation. It might just work out of the box, but that would need to be tested.

I'm mostly commenting for visibility, not something we need to cater for at the moment, as it's going to be a while. But once this has landed in Gutenberg, we may actually want to try using it in combination with the https://wordpress.org/plugins/speculation-rules/ plugin to see what happens.

westonruter replied:
Yeah, speculation rules would most likely need to be disabled on sites with full page client-side navigation. This is because each link click would have a `preventDefault` and so the normal browser navigation would never happen.  So the speculative loads would likely be wasted, especially the `prerender` ones. It may be that `prefetch` speculative loads might be able to be reused, but I'm unsure if the `fetch()` from the Interactivity API could reuse them.

felixarntz replied:
Potentially we don't have to completely disable them. There may be ways to configure the rules so that any `a` tags with those specific Interactivity API attributes would be skipped.

This wouldn't be helpful if the entire page is using them, but for pages partially using this it would give the best of both worlds. Worth trying when we actually start experimenting with the integration.

westonruter replied:
I just put together a quick test of how Speculation Rules work with client-side navigation: https://speculation-rules-with-client-side-navigation.glitch.me/

~In short, `prefetch` _does_ work but `prerender` does not.~

~I see that clicking the "with prefetch" link results in `(prefetch cache)` appearing for the entry in the network log, whereas the "with prerender" link is the same as if there was no speculative load at all.~

~All this to say, it will be important for `prefetch` to be used exclusively on client-side navigation sites. Unfortunately this means they'll miss out on instant page loads that come with prerendering (including loading the subresources like images).~

westonruter replied:
Correction: It seems prerendering _will_ add the subresources to the cache, but the prerendered HTML is not served from the prefetch cache for the `fetch()`.

westonruter replied:
Correction 2: It seems Glitch sending `max-age=0` for the speculative prerender response prevented it from being served from cache, while this wasn't the case for the speculative prefetch response. I removed `max-age=0` from that response, and now it seems speculative prerenders _are_ served from cache for client-side navigations via `fetch()`: https://speculation-rules-with-client-side-navigation.glitch.me/

SantosGuillamot replied:
I added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore deeper how both projects interact after this experiment.

felixarntz commented on lib/experimental/full-page-client-side-navigation.php:
```diff
@@ -0,0 +1,60 @@
+		if ( empty( $p->get_attribute( 'data-wp-on--click' ) ) ) {
+			$p->set_attribute( 'data-wp-on--click', 'core/router::actions.navigate' );
+		}
+		if ( empty( $p->get_attribute( 'data-wp-on--mouseenter' ) ) ) {
```
What about mobile devices? Does `mouseenter` still work there somehow? Based on the event name I would think we need to also support a mobile friendly event.

westonruter replied:
AFAIK no, `mouseenter` does not work on mobile. Closest thing would be `touchstart`.

westonruter replied:
This is also a good example for where multiple actions should be able to be specified for a given event, as we've been discussing in https://github.com/WordPress/gutenberg/discussions/60574. There should not be a requirement that the `data-wp-on--mouseenter` attribute be empty. It should rather do something like:

```php
$p->set_attribute( 'data-wp-on--mouseenter--core-router', 'core/router::actions.prefetch' );
```

felixarntz replied:
Maybe let's wait until #60661 has landed, then we can fix these attributes to be unconditionally added, using a suffix for uniqueness.

SantosGuillamot replied:
I added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore and decide how to interact with mobile events.

felixarntz commented on packages/interactivity-router/src/index.js:
```diff
@@ -128,23 +189,34 @@ export const { state, actions } = store( 'core/router', {
 		 * @return {Promise} Promise that resolves once the navigation is completed or aborted.
 		 */
-		*navigate( href, options = {} ) {
+		*navigate( eventOrUrl, options = {} ) {
```
Having `eventOrUrl` here is not great from an API perspective. Where's `navigate()` used with a URL? I would suggest we implement a separate action for that, e.g. `navigateUrl()` or `navigateToUrl()`.

This doesn't mean we have to duplicate all the code. We can probably have a helper function with the reusable bits somewhere, or maybe one action could even call the other.

SantosGuillamot replied:
In [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/d72cd023eae40645e8e787f24d91f515b12ce4ee), I changed the strategy and it registers the event handlers in JS instead of using directives. That allow us to reuse the existing navigate and prefetch functions as they are today.

felixarntz commented on packages/interactivity-router/src/index.js:
```diff
@@ -229,17 +305,26 @@ export const { state, actions } = store( 'core/router', {
+		 * @param {string}       [options.html]  HTML string to be used instead of fetching the requested URL.
 		 */
-		prefetch( url, options = {} ) {
+		prefetch( eventOrUrl, options = {} ) {
```
Same here, can we separate this into two actions, one expecting `event`, the other expecting `url`?

SantosGuillamot replied:
In [this commit](https://github.com/WordPress/gutenberg/pull/59707/commits/d72cd023eae40645e8e787f24d91f515b12ce4ee), I changed the strategy and it registers the event handlers in JS instead of using directives. That allow us to reuse the existing navigate and prefetch functions as they are today.

westonruter commented on packages/interactivity-router/src/head.js:
```diff
@@ -0,0 +1,91 @@
+	// Detect nodes that should be added or removed.
+	for ( const child of document.head.children ) {
+		const id = getTagId( child );
+		// Always remove styles and links as they might change.
+		if ( child.nodeName === 'LINK' || child.nodeName === 'STYLE' )
+			toRemove.push( child );
```
If the stylesheet remains the same, I think it should be left as-is. Otherwise, removing a stylesheet and then adding it right back in will likely cause a performance problem.

Example: https://mousy-citrine-lotus.glitch.me/

Note there is a flash of unstyled content. It also causes layout and style recalculation:

![image](https://github.com/WordPress/gutenberg/assets/134745/d4b8c70f-28d2-409b-bdff-9a43862f7e67)

westonruter replied:
Would be nice to be able to reuse DOM diffing from Preact here. (Not sure if that's possible.)

SantosGuillamot replied:
I added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore this after this experiment is merged.

luisherranz replied:
I tried using Preact at first, but if the scripts and styles change order or new elements appear in the head, Preact gets confused and deletes and re-adds everything.

I think it will be simpler and more stable if we control it ourselves.

westonruter commented on packages/interactivity-router/src/index.js:
```diff
@@ -102,12 +134,41 @@ window.addEventListener( 'popstate', async () => {
 );
 
+// Check if the link is valid for client-side navigation.
+const isValidLink = ( ref ) =>
+	ref &&
+	ref instanceof window.HTMLAnchorElement &&
+	ref.href &&
+	( ! ref.target || ref.target === '_self' ) &&
+	ref.origin === window.location.origin;
```
Also need to exclude `wp-admin` links. See similar logic in Customizer preview: https://github.com/WordPress/wordpress-develop/blob/f525e665b6ea6e88289d82cad5fc1dfac57db109/src/js/_enqueues/wp/customize/preview.js#L283-L323

felixarntz replied:
Also `wp-login.php`. See similar [Speculative Loading plugin exclusions](https://github.com/WordPress/performance/blob/trunk/plugins/speculation-rules/helper.php#L43).

In fact, we're even providing a [filter](https://github.com/WordPress/performance/blob/trunk/plugins/speculation-rules/helper.php#L62) (though that could be added later here).

westonruter replied:
Also important to check for whether the link is for an anchor link for another part of the page: https://github.com/WordPress/wordpress-develop/blob/f525e665b6ea6e88289d82cad5fc1dfac57db109/src/js/_enqueues/wp/customize/preview.js#L156-L160

So this here should also check if `! ref.getAttribute('href').startsWith('#')`.

SantosGuillamot replied:
Thanks for the feedback and the links! I added the checks for `wp-admin`, `wp-login`, `_wpnonce`, and anchors as part of [this conditional](https://github.com/WordPress/gutenberg/pull/59707/files#diff-6b0e711ee1a7c83c535664372c8a0db6b2025cb33e9263bab915dff5145e01e7R164-R167). Let me know if that's not what you had in mind.

SantosGuillamot committed:
Add Promise to JSDoc

Co-authored-by: Michal <mmczaplinski@gmail.com>

SantosGuillamot commented:
Thanks a lot for all the feedback. I've been reviewing all the remaining comments and I believe this is how I will proceed:

### To do in this pull request
- Use event delegation on the client instead of using `data-wp-on--click` directive in every link: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1541233377). 
- Previous task will allow us to use just one `navigate` function with the existing format, using `url` as the parameter: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1559669125).
- Accept only a string in the `url` parameter and adapt the tests to reflect that: [link](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2022343602).
- Use a static variable to add the `<body>` tag once: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1548155129). And potentially use a closing tag.
- Exclude `wp-admin` and `wp-login` from links and ensure anchors work as expected: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1560011157).
- Wrap fetch in a try/catch (the whole conditional): [link](https://github.com/WordPress/gutenberg/pull/59707/files/743926f0160e985a4641c3b158f64700e8c656ea#r1548734375).
- Add experimental message in the help attribute: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1548023233).
- Rename document parameter: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1548730074).
- Check that the code runs in the Gutenberg plugin: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1540909210).

### To improve after that
- Explore deeper how to use event delegation in the client: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1541233377).
- Add e2e tests: [link](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2031730825).
- Explore how to update query attribute when the experiment is enabled: [link](https://github.com/WordPress/gutenberg/pull/59707#pullrequestreview-1974040313). It looks like a core bug.
- Revisit how to fetch head assets: [link](https://github.com/WordPress/gutenberg/pull/59707#pullrequestreview-1975274682).
    - Change innerText for textContent and revisit how styles are injected: [link](https://github.com/WordPress/gutenberg/pull/59707).
    - Check if **[Preact fails to reconcile scripts correctly when doing client-side navigation](https://github.com/WordPress/block-interactivity-experiments/issues/242)** is still an issue.
    - If the stylesheet remains the same, it should probably be left as-is: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1559964017). It would be nice to be able to reuse DOM diffing from Preact here if possible.
- Explore how to modify the body tag without the hack: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1559952673).
- Explore how this interacts with Speculation Rules API: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1559663664).
- Explore if we want to add event listeners to mobile: [link](https://github.com/WordPress/gutenberg/pull/59707#discussion_r1559664794). (Maybe on click is enough)

SantosGuillamot committed:
Wrap fullPage code in IS_GUTENBERG_PLUGIN check

SantosGuillamot committed:
Rename document variable to doc

SantosGuillamot committed:
Use static variable to add body directive once

SantosGuillamot committed:
Wrap fetch in try/catch

SantosGuillamot committed:
Specify experimental in query help message

SantosGuillamot committed:
Prevent client navigation in admin links

SantosGuillamot committed:
Add event listeners for navigate and prefetch in JS

SantosGuillamot committed:
Add check for anchor links of the same page

SantosGuillamot commented:
I believe this experiment is ready to be merged. I addressed all the feedback listed [here](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203). We should probably create a tracking issue for client-side navigation and add the follow-up tasks there, and potentially create specific issues for some of them.

SantosGuillamot commented:
I believe this experiment is ready to be merged. I addressed all the feedback listed [here](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203). We should probably create a tracking issue for client-side navigation and add the follow-up tasks there, and potentially create specific issues for some of them.

westonruter commented on packages/interactivity-router/src/index.js:
```diff
@@ -102,12 +135,47 @@ window.addEventListener( 'popstate', async () => {
+	ref.href &&
+	( ! ref.target || ref.target === '_self' ) &&
+	ref.origin === window.location.origin &&
+	! ref.pathname.startsWith( '/wp-admin' ) &&
+	! ref.pathname.startsWith( '/wp-login.php' ) &&
```
Sometimes WordPress sites install core in a subdirectory, like `/wp/`. In such cases `home_url()` and `site_url()` are different, e.g. `http://example.com/` vs `https://example.com/wp/`. 

Additionally, in multisite subdirectory installs, there will also be path segment(s) before `/wp-admin`. 

Easiest way to handle that right now is to just check if the path includes the strings:

```suggestion
	! ref.pathname.includes( '/wp-admin' ) &&
	! ref.pathname.includes( '/wp-login.php' ) &&
```

This is not as robust as possible. But better to err on the side of not doing client-side navigation than doing it with an invalid target.

poof86 replied:
I’m simply reading the discussion, but seeing this solution would not work for my sites, because for security I always change the default wp-admin url. Maybe check for a class name or other attribute in the dom instead?

westonruter replied:
Actually, these paths should be passed from PHP so they don't have to be hard coded in JS.

SantosGuillamot replied:
You're right. Maybe we can use [wp_interactivity_config](https://developer.wordpress.org/reference/functions/wp_interactivity_config/) to create a new `isAdmin` property that is passed to the client reading [is_admin](https://developer.wordpress.org/reference/functions/is_admin/) in PHP.

cc: @DAreRodz Worth adding this to the tracking issue as a follow-up task?

westonruter replied:
I don't think it would be `is_admin()` since this would always just be `false` in the frontend context. I think we'd need to instead get an array of the special paths, including `get_admin_url()`, `wp_login_url()`, etc.

westonruter replied:
I've added this as a task on the tracking issue.

westonruter commented on packages/interactivity-router/src/index.js:
```diff
@@ -246,3 +317,34 @@ export const { state, actions } = store( 'core/router', {
+		);
+		// Prefetch on hover.
+		document.addEventListener(
+			'mouseenter',
```
On touch interfaces, I understand that the `mouseenter` event does fire but after a 300 ms delay. To ensure that prefetch happens immediately on a touch device, I suggest adding the same handler for the `touchstart` event. Or apparently `pointerenter` will handle both of these and it is supported by all browsers: https://caniuse.com/mdn-api_element_pointerenter_event

westonruter commented on packages/interactivity-router/src/index.js:
```diff
@@ -102,12 +135,47 @@ window.addEventListener( 'popstate', async () => {
+// Once this code is tested and more mature, the head should be updated for region based navigation as well.
+if ( process.env.IS_GUTENBERG_PLUGIN ) {
+	if ( navigationMode === 'fullPage' ) {
+		// Cache the scripts. Has to be called before fetching the assets.
```
Why is the caching required? Is this to guarantee execution order of the scripts when the page is loaded? What if a script is not cacheable?

westonruter replied:
Or is this rather to ensure that there is no flash of unrendered content?

michalczaplinski replied:
I believe that the original idea was simply that when doing a full-page navigation to another page, we don't want to re-download the scripts that were already present on the initial page.

If a script is not cacheable (e.g. if some of the script's contents are generated dynamically) that's indeed currently a problem. 

One solution could be examine the HTTP headers and e.g. skip the ones with `Cache-Control: no-store` , `Cache-Control: no-cache` or `Expires: <Date in the past>`

DAreRodz replied:
Thanks, @michalczaplinski, for chiming in. 😄

We want to revisit the whole assets system, and examining HTTP headers is something we could try.

BTW, I've created a tracking issue to continue with the development. https://github.com/WordPress/gutenberg/issues/60951

michalczaplinski replied:
Great! Indeed, we should take a holistic view of this. Thanks for creating the issue! 🙏

DAreRodz commented:
Thanks everyone for your work and feedback. 🙌 

I've created this [tracking issue](https://github.com/WordPress/gutenberg/issues/60951) to keep track of the progress and continue working on this feature. The list of tasks is based on all the feedback gathered from this PR.

luisherranz commented:
> I'm a little concerned about fetching scripts in fetchHeadAssets() without taking into account the script order.

@michalczaplinski We shouldn't really load unknown scripts. We just did it to get it working with the WP movies demo. 

We should only load the module scripts that belong to the new blocks on the page or other module scripts that have declared their compatibility with full-page client-side navigation.

luisherranz commented:
I arrive a bit late to the party, but I simply want to say incredible work here. Thanks to Mario for leading this and to the rest for all the valuable feedback.This is very exciting!! 👏😄
